pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,314 @@
+	ctrlerror "github.com/gardener/gardener/extensions/pkg/controller/error"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/go-logr/logr"
+	"github.com/gophercloud/gophercloud"
+	computerfip "github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/floatingips"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/layer3/floatingips"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/security/groups"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/security/rules"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// bastionEndpoints collects the endpoints the bastion host provides; the
+// private endpoint is important for opening a port on the worker node
+// ingress NSG rule to allow SSH from that node, the public endpoint is where
+// the enduser connects to establish the SSH connection.
+type bastionEndpoints struct {
+	private *corev1.LoadBalancerIngress
+	public  *corev1.LoadBalancerIngress
+}
+
+// Ready returns true if both public and private interfaces each have either
+// an IP or a hostname or both.
+func (be *bastionEndpoints) Ready() bool {
+	return be != nil && IngressReady(be.private) && IngressReady(be.public)
+}
+
+func (a *actuator) Reconcile(ctx context.Context, bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) error {
+	logger := a.logger.WithValues("bastion", client.ObjectKeyFromObject(bastion), "operation", "reconcile")
+
+	opt, err := DetermineOptions(bastion, cluster)
+	if err != nil {
+		return err
+	}
+
+	credentials, err := openstack.GetCredentials(ctx, a.client, opt.SecretReference, false)
+	if err != nil {
+		return fmt.Errorf("could not get Openstack credentials: %w", err)
+	}
+
+	openstackClientFactory, err := a.openstackClientFactory.NewFactory(credentials)
+	if err != nil {
+		return fmt.Errorf("could not create Openstack client factory: %w", err)
+	}
+
+	securityGroup, err := ensureSecurityGroup(openstackClientFactory, opt)
+	if err != nil {
+		return err
+	}
+
+	err = ensureSecurityGroupRules(openstackClientFactory, opt, securityGroup.ID)
+	if err != nil {
+		return err
+	}
+
+	instance, err := ensureComputeInstance(logger, openstackClientFactory, opt)
+	if err != nil || instance == nil {
+		return err
+	}
+
+	fipid, err := ensurePublicIPAddress(opt, openstackClientFactory)
+	if err != nil {
+		return err
+	}
+
+	ready, err := ensureAssociateFIPWithInstance(openstackClientFactory, instance, fipid)
+	if err != nil || !ready {
+		return err
+	}
+
+	// refresh instance after public ip attached/created
+	instancenstatus, err := getBastionInstance(openstackClientFactory, opt.BastionInstanceName)
+	if err != nil {
+		return err
+	}
+
+	// check if the instance already exists and has an IP
+	endpoints, err := getInstanceEndpoints(&instancenstatus[0], opt)
```

Comment:
won't this segfault if `getBastionInstance` finds no results ?

Reply:
* before `getInstanceEndpoints` there is a `ensureComputeInstance` function ensure instance exist.  https://github.com/tedteng/gardener-extension-provider-openstack/blob/c48a28149d19761c41135b2ec86de664c58f1bc5/pkg/controller/bastion/actuator_reconcile.go#L83-L106

* inside the `getInstanceEndpoints` function will also check if the instance is nil  https://github.com/tedteng/gardener-extension-provider-openstack/blob/c48a28149d19761c41135b2ec86de664c58f1bc5/pkg/controller/bastion/actuator_reconcile.go#L196-L199

```
	instance, err := ensureComputeInstance(logger, openstackClientFactory, opt)
        ....

	// refresh instance after public ip attached/created
	instancenstatus, err := getBastionInstance(openstackClientFactory, opt.BastionInstanceName)
	if err != nil {
		return err
	}

	// check if the instance already exists and has an IP
	endpoints, err := getInstanceEndpoints(&instancenstatus[0], opt)
```


```
  func getInstanceEndpoints(instance *servers.Server, opt *Options) (*bastionEndpoints, error) {
	  if instance == nil || instance.Status != "ACTIVE" {
		  return nil, fmt.Errorf("compute instance can't be nil")
	  }
```

Reply:
This is maybe a bit pedantic but consider the following: If between the `ensureInstance` and `getInstanceEndpoints` the instance is deleted (from whatever random event) the result will be that the pod will crash and you checks will not help since the `instancestatus` list will be empty and the 1st element can't be derefenced.

Reply:
committed in 222d9eb1